# CEREsFit Linear Regression - Example

This is an example notebook to show the usage of the Mahon linear regression package.
It is expected that you have installed the `ceresfit`.
To do so, run:

```
pip install ceresfit
```

In [12]:
from ceresfit import LinReg
import numpy as np

## Data preparation

First, let us prepare some data. This data set is identical to data set 1 in Stephan and Trappitsch (2022).

In [13]:
xdat = np.array([0.0370, 0.0350, 0.0320, 0.0400, 0.0130, 0.0380, 0.0420, 0.0300])  # x data
xunc = xdat * 0.03  # x uncertainty
ydat = np.array([0.00080, 0.00084, 0.00100, 0.00085, 0.00270, 0.00071, 0.00043, 0.00160])  # y data
yunc = ydat * 0.1  # y uncertainty

# correlation factors for uncertainty
rho = np.zeros_like(xdat) + 0.707106781186548

## Linear Regression with Correlated Uncertainties

This calculation is equal to case 1a in the paper by Stephan and Trappitsch (2022). 

In [14]:
case_a = LinReg(xdat, xunc, ydat, yunc, rho=rho)

We can now display the parameters and their uncertainties as following:

In [15]:
case_a.slope

(-0.07751574256394125, 0.010270918045982397)

In [16]:
case_a.intercept

(0.0036841599044688195, 0.0003752339783879849)

In [17]:
case_a.mswd

1.0443169785938846

In [18]:
case_a.chi_squared

6.265901871563308

Alternatively, all parameters can be displayed at once in the order
slope, slope uncertainty, intercept, intercept uncertainty, MSWD:

In [19]:
case_a.parameters

array([-7.75157426e-02,  1.02709180e-02,  3.68415990e-03,  3.75233978e-04,
        1.04431698e+00])

An interesting aspect that users might want to check are the 95% confidence intervals that you would expect the MSWD value to lie in. This depends on the degrees of freedom of your system. The `LinReg` class contains a method to calculate the lower and upper bounds of the 95% confidence interval for the MSWD.

In [20]:
case_a.mswd_ci()

(0.2062240409652005, 2.4082292225746533)

The above calculate MSWD of 1.04 lies perfectly well within this interval.

## Linear Regression with Correlated Uncertainties

This is equivalent to case 1b in the paper. Here, we can simply leave the factor `rho` away.

In [21]:
case_b = LinReg(xdat, xunc, ydat, yunc)
case_b.parameters

array([-7.63850005e-02,  9.51997663e-03,  3.64110926e-03,  3.49027429e-04,
        1.73032126e+00])

## Fixed point linear regression

The package also provides the possibility to calculate a fixed point linear regression (case 1c). To do so, we first have to define a fixed point and then add it as an argument to the linear regression.

In [22]:
fix_point = np.array([0.01, 0.003])

case_c = LinReg(xdat, xunc, ydat, yunc, rho=rho, fixpt=fix_point)
case_c.parameters

array([-8.08426056e-02,  2.22210172e-03,  3.80842606e-03,  2.22210172e-05,
        9.11015003e-01])